expression/builtin_regexp.go

```diff
@@ -0,0 +1,1281 @@
+		if err != nil {
+			return err
+		}
+
+		// Find string
+		if re.isBinCollation {
+			matches := reg.FindAll(bexpr, -1)
+			length := int64(len(matches))
+			if length == 0 || occurrence > length {
+				result.AppendNull()
+				continue
+			}
+
+			result.AppendString(fmt.Sprintf("0x%s", strings.ToUpper(hex.EncodeToString(matches[occurrence-1]))))
+		} else {
+			matches := reg.FindAllString(expr, -1)
+			length := int64(len(matches))
+			if length == 0 || occurrence > length {
+				result.AppendNull()
+				continue
+			}
+
+			result.AppendString(matches[occurrence-1])
+		}
+	}
+	return nil
+}
+
+// ---------------------------------- regexp_instr ----------------------------------
+
+type regexpInStrFunctionClass struct {
+	baseFunctionClass
+}
+
+func (c *regexpInStrFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (builtinFunc, error) {
+	if err := c.verifyArgs(args); err != nil {
+		return nil, err
+	}
+
+	argTp := []types.EvalType{types.ETString, types.ETString}
+	switch len(args) {
+	case 3:
+		argTp = append(argTp, types.ETInt)
+	case 4:
+		argTp = append(argTp, types.ETInt, types.ETInt)
+	case 5:
+		argTp = append(argTp, types.ETInt, types.ETInt, types.ETString)
+	case 6:
+		argTp = append(argTp, types.ETInt, types.ETInt, types.ETInt, types.ETString)
+	}
+
+	bf, err := newBaseBuiltinFuncWithTp(ctx, c.funcName, args, types.ETInt, argTp...)
+	if err != nil {
+		return nil, err
+	}
+
+	bf.tp.SetFlen(mysql.MaxIntWidth)
+	sig := regexpInStrFuncSig{
+		regexpBaseFuncSig: regexpBaseFuncSig{baseBuiltinFunc: bf},
+	}
+
+	if bf.collation == charset.CollationBin {
+		sig.setPbCode(tipb.ScalarFuncSig_RegexpInStrSig)
+		sig.isBinCollation = true
+	} else {
+		sig.setPbCode(tipb.ScalarFuncSig_RegexpInStrUTF8Sig)
+		sig.isBinCollation = false
+	}
+
+	return &sig, nil
+}
+
+type regexpInStrFuncSig struct {
+	regexpBaseFuncSig
+	isBinCollation bool
+}
+
+func (re *regexpInStrFuncSig) Clone() builtinFunc {
+	newSig := &regexpLikeFuncSig{}
+	newSig.cloneFrom(&re.baseBuiltinFunc)
+	newSig.clone(&re.regexpBaseFuncSig)
+	return newSig
+}
+
+func (re *regexpInStrFuncSig) vectorized() bool {
+	return true
+}
+
+// Call this function when all args are constant
+func (re *regexpInStrFuncSig) evalInt(row chunk.Row) (int64, bool, error) {
+	expr, isNull, err := re.args[0].EvalString(re.ctx, row)
+	if isNull || err != nil {
+		return 0, true, err
+	}
+
+	pat, isNull, err := re.args[1].EvalString(re.ctx, row)
+	if isNull || err != nil {
+		return 0, true, err
+	}
+
+	pos := int64(1)
+	occurrence := int64(1)
+	returnOption := int64(0)
+	matchType := ""
+	arg_num := len(re.args)
+	var bexpr []byte
+
+	if re.isBinCollation {
+		bexpr = []byte(expr)
+	}
+
+	if arg_num >= 3 {
+		pos, isNull, err = re.args[2].EvalInt(re.ctx, row)
+		if isNull || err != nil {
+			return 0, true, err
+		}
+
+		// Check position and trim expr
+		if re.isBinCollation {
+			if pos < 1 || pos > int64(len(bexpr)) {
+				if len(expr) != 0 || (len(expr) == 0 && pos != 1) {
```

Comment:
looks like `pos > int64(len(bexpr))` does not always return invalid index here? Is it expected?

Reply:
> looks like `pos > int64(len(bexpr))` does not always return invalid index here? Is it expected?

Yes, when expr = "" and pos = 1, pos is greater than len(expr), but it's a valid index.